### PR TITLE
Create sets from cards

### PR DESCRIPTION
**Pull request by hannahbishop**
Prior to this commit set objects were created with a length and value.
Creating a set from a list of cards is a better model for what a set
represents.

### CODE REVIEW

File: lib/pokerscore/set.rb [hunk 1]
```diff
@@ -1,24 +1,36 @@
...
+  private
+
+  def cards_same_value?
+    @cards.map(&:value).uniq.size == 1
+  end
+
+  def set_size
+    @cards.size
   end
 end
```
Comment (jgayfer) [inline lib/pokerscore/set.rb]: Beautiful 👌

File: lib/pokerscore/pair_search.rb [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+  class PairSearch
+
+    def call(hand)
+      pairs = []
+      grouped_cards_by_value = hand.cards.group_by(&:value)
+      grouped_cards_by_value.each_value do |cards|
+        pairs << PokerScore::SetFactories::pair(cards) if cards.size == 2
+      end
+      pairs
+    end
```
Comment (jgayfer) [inline lib/pokerscore/pair_search.rb]: We might be able to get away with dropping the `grouped_cards_by_value` variable here?
```
hand.cards.group_by(&:value) do |cards_with_same_value|
  pairs << PokerScore::SetFactories::pair(cards_with_same_value) if cards.size == 2
end
```

We _could_ also get fancy with a `tap` here, though last time I did this my team members were up in arms (though I still think it's worth using, as I don't like having to explicitly return `pairs`)

```
def call(hand)
  [].tap do |pairs|
    hand.cards.group_by(&:value) do |cards_with_same_value|
      pairs << PokerScore::SetFactories::pair(cards_with_same_value) if cards.size == 2
    end
  end
end
```

This is mostly syntactic sugar though, and is somewhat contentious (at least to Dennis and Jon)

Comment (hannahbishop) [reply lib/pokerscore/pair_search.rb]: Addressed in person, and has been cleaned up

**Comment by swcraig** [comment]
:+1:

**Pull request merged**